File: Minimal/src/Brontide.Minimal.Binding/Portable/PortableLifecycle.fs
namespace Brontide.Minimal.Binding.Portable

/// The Channel envelope kinds.
///
/// There is deliberately no denial kind: a local authority denial emits no frame at all, so the
/// baseline experiment's denial message is not carried into the portable envelope set. Process loss
/// is likewise observed locally rather than signalled.
[<RequireQualifiedAccess>]
type EnvelopeKind =
    | Establish
    | EstablishAccepted
    | Ready
    | Request
    | Outcome
    | ProtocolError
    | Withdraw
    | Terminate

[<RequireQualifiedAccess>]
module EnvelopeKind =
    let token kind =
        match kind with
        | EnvelopeKind.Establish -> "establish"
        | EnvelopeKind.EstablishAccepted -> "establish-accepted"
        | EnvelopeKind.Ready -> "ready"
        | EnvelopeKind.Request -> "request"
        | EnvelopeKind.Outcome -> "outcome"
        | EnvelopeKind.ProtocolError -> "protocol-error"
        | EnvelopeKind.Withdraw -> "withdraw"
        | EnvelopeKind.Terminate -> "terminate"

    let all =
        [ EnvelopeKind.Establish
          EnvelopeKind.EstablishAccepted
          EnvelopeKind.Ready
          EnvelopeKind.Request
          EnvelopeKind.Outcome
          EnvelopeKind.ProtocolError
          EnvelopeKind.Withdraw
          EnvelopeKind.Terminate ]

    let tryParse value : PortableResult<EnvelopeKind> =
        match all |> List.tryFind (fun kind -> token kind = value) with
        | Some kind -> Ok kind
        | None ->
            refuse
                ProtocolCategory.UnsupportedKind
                "unknown-kind"
                $"Envelope kind '{value}' is outside the declared set; an unknown kind is never treated as a no-op."

    /// A request and an outcome both require a request identity; neither is ever matched to an
    /// outstanding request by position.
    let requiresRequestId kind =
        kind = EnvelopeKind.Request || kind = EnvelopeKind.Outcome

/// The explicit binding lifecycle states.
///
/// The baseline experiments left their lifecycles implicit in message ordering. Making the states
/// explicit is what lets an illegal transition be a state violation rather than an accident of
/// arrival order.
[<RequireQualifiedAccess>]
type LifecycleState =
    | Unestablished
    | Establishing
    | Established
    | Ready
    | Active
    | Withdrawn
    | Terminated
    | Failed

[<RequireQualifiedAccess>]
module LifecycleState =
    let token state =
        match state with
        | LifecycleState.Unestablished -> "unestablished"
        | LifecycleState.Establishing -> "establishing"
        | LifecycleState.Established -> "established"
        | LifecycleState.Ready -> "ready"
        | LifecycleState.Active -> "active"
        | LifecycleState.Withdrawn -> "withdrawn"
        | LifecycleState.Terminated -> "terminated"
        | LifecycleState.Failed -> "failed"

    let isTerminal state =
        state = LifecycleState.Terminated || state = LifecycleState.Failed

/// The lifecycle of one binding scope at one endpoint, together with the accepted request
/// identities inside the declared replay window.
///
/// The record is immutable and every transition returns a new one, so an illegal transition leaves
/// the previous state untouched rather than half-applied.
[<StructuralEquality; NoComparison>]
type Lifecycle =
    private
        { State: LifecycleState
          AcceptedRequests: Set<string>
          ReplayProtectionDeclared: bool }

[<RequireQualifiedAccess>]
module Lifecycle =

    let private transitions =
        [ LifecycleState.Unestablished, EnvelopeKind.Establish, LifecycleState.Establishing
          LifecycleState.Establishing, EnvelopeKind.EstablishAccepted, LifecycleState.Established
          LifecycleState.Established, EnvelopeKind.Ready, LifecycleState.Ready
          LifecycleState.Ready, EnvelopeKind.Request, LifecycleState.Active
          LifecycleState.Active, EnvelopeKind.Outcome, LifecycleState.Ready
          LifecycleState.Ready, EnvelopeKind.Withdraw, LifecycleState.Withdrawn
          LifecycleState.Active, EnvelopeKind.Withdraw, LifecycleState.Withdrawn
          LifecycleState.Withdrawn, EnvelopeKind.Terminate, LifecycleState.Terminated
          LifecycleState.Ready, EnvelopeKind.Terminate, LifecycleState.Terminated
          LifecycleState.Established, EnvelopeKind.Terminate, LifecycleState.Terminated ]

    let create replayProtectionDeclared =
        { State = LifecycleState.Unestablished
          AcceptedRequests = Set.empty
          ReplayProtectionDeclared = replayProtectionDeclared }

    let state (lifecycle: Lifecycle) = lifecycle.State

    /// Adopts the replay declaration the Binding Plan froze. Establishment is the only point at
    /// which the window can start.
    let declareReplayProtection declared (lifecycle: Lifecycle) =
        { lifecycle with ReplayProtectionDeclared = declared }

    let fail (lifecycle: Lifecycle) =
        { lifecycle with State = LifecycleState.Failed }

    /// Applies one envelope kind, or refuses it as a state violation.
    let apply kind (lifecycle: Lifecycle) : PortableResult<Lifecycle> =
        if LifecycleState.isTerminal lifecycle.State then
            stateViolation
                "terminal-state"
                $"No frame is legal in terminal state '{LifecycleState.token lifecycle.State}'; '{EnvelopeKind.token kind}' was refused."
        elif kind = EnvelopeKind.ProtocolError then
            // A protocol error is legal in every non-terminal state and always ends the binding.
            Ok { lifecycle with State = LifecycleState.Failed }
        else
            match
                transitions
                |> List.tryFind (fun (from, on, _) -> from = lifecycle.State && on = kind)
            with
            | Some(_, _, next) -> Ok { lifecycle with State = next }
            | None ->
                stateViolation
                    "illegal-transition"
                    $"A '{EnvelopeKind.token kind}' frame is illegal in state '{LifecycleState.token lifecycle.State}'."

    /// Records an accepted request identity inside the declared replay window, refusing a repeat.
    let recordRequest (request: ChannelRequestId) (lifecycle: Lifecycle) : PortableResult<Lifecycle> =
        if not lifecycle.ReplayProtectionDeclared then
            Ok lifecycle
        else
            let identity = ChannelRequestId.value request

            if Set.contains identity lifecycle.AcceptedRequests then
                refuse
                    ProtocolCategory.ReplayDetected
                    "replay"
                    $"Request '{identity}' already appeared inside the declared replay window."
            else
                Ok
                    { lifecycle with
                        AcceptedRequests = Set.add identity lifecycle.AcceptedRequests }

File: Reference/src/Brontide.Reference.Experimental.Binding/Portable/PortableLifecycle.cs
using System.Collections.Immutable;

namespace Brontide.Reference.Experimental.Binding.Portable;

/// <summary>The explicit binding lifecycle states.</summary>
/// <remarks>
/// The baseline experiments left their lifecycles implicit in message ordering. Making the states
/// explicit is what lets an illegal transition be a state violation rather than an accident of
/// arrival order.
/// </remarks>
public enum PortableLifecycleState
{
    Unestablished,
    Establishing,
    Established,
    Ready,
    Active,
    Withdrawn,
    Terminated,
    Failed
}

public static class PortableLifecycleStates
{
    public static string Token(this PortableLifecycleState state) => state switch
    {
        PortableLifecycleState.Unestablished => "unestablished",
        PortableLifecycleState.Establishing => "establishing",
        PortableLifecycleState.Established => "established",
        PortableLifecycleState.Ready => "ready",
        PortableLifecycleState.Active => "active",
        PortableLifecycleState.Withdrawn => "withdrawn",
        PortableLifecycleState.Terminated => "terminated",
        PortableLifecycleState.Failed => "failed",
        _ => throw new ArgumentOutOfRangeException(nameof(state))
    };

    public static bool IsTerminal(this PortableLifecycleState state) =>
        state is PortableLifecycleState.Terminated or PortableLifecycleState.Failed;
}

/// <summary>
/// The lifecycle state machine for one binding scope at one endpoint.
/// </summary>
/// <remarks>
/// No provider effect occurs before the ready state is reached, a request is legal only in ready
/// because the declared concurrency is one, and correlation mismatch, illegal state, peer
/// termination, timeout, and interruption never fabricate a success.
/// </remarks>
public sealed class PortableLifecycle
{
    private static readonly ImmutableArray<(PortableLifecycleState From, PortableEnvelopeKind On, PortableLifecycleState To)> Transitions =
    [
        (PortableLifecycleState.Unestablished, PortableEnvelopeKind.Establish, PortableLifecycleState.Establishing),
        (PortableLifecycleState.Establishing, PortableEnvelopeKind.EstablishAccepted, PortableLifecycleState.Established),
        (PortableLifecycleState.Established, PortableEnvelopeKind.Ready, PortableLifecycleState.Ready),
        (PortableLifecycleState.Ready, PortableEnvelopeKind.Request, PortableLifecycleState.Active),
        (PortableLifecycleState.Active, PortableEnvelopeKind.Outcome, PortableLifecycleState.Ready),
        (PortableLifecycleState.Ready, PortableEnvelopeKind.Withdraw, PortableLifecycleState.Withdrawn),
        (PortableLifecycleState.Active, PortableEnvelopeKind.Withdraw, PortableLifecycleState.Withdrawn),
        (PortableLifecycleState.Withdrawn, PortableEnvelopeKind.Terminate, PortableLifecycleState.Terminated),
        (PortableLifecycleState.Ready, PortableEnvelopeKind.Terminate, PortableLifecycleState.Terminated),
        (PortableLifecycleState.Established, PortableEnvelopeKind.Terminate, PortableLifecycleState.Terminated)
    ];

    private readonly HashSet<string> _acceptedRequests = new(StringComparer.Ordinal);
    private bool _replayProtectionDeclared;

    public PortableLifecycle(bool replayProtectionDeclared) => _replayProtectionDeclared = replayProtectionDeclared;

    /// <summary>
    /// Adopts the replay declaration the Binding Plan froze. It is declared during establishment,
    /// which is the only point at which the window can start.
    /// </summary>
    public void DeclareReplayProtection(bool declared) => _replayProtectionDeclared = declared;

    public PortableLifecycleState State { get; private set; } = PortableLifecycleState.Unestablished;

    /// <summary>Applies one envelope kind, or refuses it as a state violation.</summary>
    public void Apply(PortableEnvelopeKind kind)
    {
        // A protocol error is legal in every non-terminal state and always ends the binding.
        if (kind == PortableEnvelopeKind.ProtocolError)
        {
            RequireNotTerminal(kind);
            State = PortableLifecycleState.Failed;
            return;
        }

        RequireNotTerminal(kind);
        foreach (var transition in Transitions)
        {
            if (transition.From == State && transition.On == kind)
            {
                State = transition.To;
                return;
            }
        }

        throw new PortableFaultException(
            PortableProtocolCategory.StateViolation,
            "illegal-transition",
            $"A '{kind.Token()}' frame is illegal in state '{State.Token()}'.");
    }

    public void Fail() => State = PortableLifecycleState.Failed;

    /// <summary>
    /// Records an accepted request identity inside the declared replay window, refusing a repeat.
    /// </summary>
    public void RecordRequest(PortableChannelRequestId request)
    {
        if (!_replayProtectionDeclared)
        {
            return;
        }

        if (!_acceptedRequests.Add(request.Value))
        {
            throw new PortableFaultException(
                PortableProtocolCategory.ReplayDetected,
                "replay",
                $"Request '{request}' already appeared inside the declared replay window.");
        }
    }

    private void RequireNotTerminal(PortableEnvelopeKind kind)
    {
        if (State.IsTerminal())
        {
            throw new PortableFaultException(
                PortableProtocolCategory.StateViolation,
                "terminal-state",
                $"No frame is legal in terminal state '{State.Token()}'; '{kind.Token()}' was refused.");
        }
    }
}
